zilliqa/src/consensus.rs

```diff
@@ -832,83 +881,158 @@ impl Consensus {
                 vote_view = block_view + 1,
                 "storing vote"
             );
+            self.votes.insert(
+                block_hash,
+                (
+                    signatures.clone(),
+                    cosigned.clone(),
+                    cosigned_weight,
+                    supermajority_reached,
+                ),
+            );
             if supermajority_reached {
                 // if we are already in the round in which the vote counts and have reached supermajority
                 if block_view + 1 == self.view.get_view() {
-                    let qc =
-                        self.qc_from_bits(block_hash, &signatures, cosigned.clone(), block_view);
-                    let parent_hash = qc.block_hash;
-                    let parent = self
-                        .get_block(&parent_hash)?
-                        .ok_or_else(|| anyhow!("missing block"))?;
-                    let parent_header = parent.header;
+                    // We propose new block immediately if there's something in mempool or it's the first view
+                    // Otherwise the block will be proposed on timeout
 
-                    let previous_state_root_hash = self.state.root_hash()?;
+                    let transactions_count = self.transaction_pool.size();
 
-                    if previous_state_root_hash != parent.state_root_hash() {
-                        warn!("when proposing, state root hash mismatch, expected: {:?}, actual: {:?}", parent.state_root_hash(), previous_state_root_hash);
-                        self.state.set_to_root(H256(parent.state_root_hash().0));
+                    if (self.view.get_view() == 1)
+                        || (block_view + 1 == self.view.get_view() && transactions_count > 0)
+                    {
+                        return self.propose_new_block();
+                    } else {
+                        // Check if there's enough time to wait on a timeout and then propagate an empty block in the network before other participants trigger NewView
+                        let (time_since_last_view_change, exponential_backoff_timeout) =
+                            self.get_consensus_timeout_params();
+                        let minimum_time_left_for_empty_block =
```

Comment:
Wouldn't it make sense if `minimum_time_left_for_empty_block()` also returned the `minimum_time_left_for_empty_block`?